File: src/sbbs3/qtmonitor/mqtt_client.py
"""MQTT client wrapper using paho-mqtt, bridged to Qt signals."""

import paho.mqtt.client as paho_mqtt
import paho.mqtt.properties as mqtt_properties
from PySide6.QtCore import QObject, Signal, Slot, Qt
from tls_psk import make_psk_context

# broker.js sends SubscriptionIdentifier=0 when no subscription ID was
# requested — valid intent but violates MQTT 5.0 spec (must be 1+).
# Patch paho's validation to accept 0.
_orig_setattr = mqtt_properties.Properties.__setattr__
def _patched_setattr(self, name, value):
    if name == "SubscriptionIdentifier" and value == 0:
        object.__setattr__(self, name, value)
        return
    _orig_setattr(self, name, value)
mqtt_properties.Properties.__setattr__ = _patched_setattr


class MqttClient(QObject):
    """Wraps paho-mqtt with Qt signal-based topic routing.

    Paho runs its network loop on a background thread; all signals are
    emitted via queued connections so slots execute on the Qt main thread.
    """

    connected = Signal()
    disconnected = Signal()
    error_occurred = Signal(str)

    log_message = Signal(str, int, str, str)  # server, level, timestamp, text
    node_status = Signal(int, dict)            # node_num, fields
    node_verbose = Signal(int, str)            # node_num, description
    client_update = Signal(str, str, dict)     # server, action, fields
    server_state = Signal(str, str)            # server, state
    server_stat = Signal(str, str, str)        # server, stat_name, value
    login_attempt = Signal(str, str, dict)     # ip, action, fields

    _emit_connected = Signal()
    _emit_disconnected = Signal()
    _emit_error = Signal(str)
    _emit_message = Signal(str, bytes)

    def __init__(self, parent=None):
        super().__init__(parent)
        self._bbs_id = None
        self._host = "localhost"
        self._port = 1883
        self._username = None
        self._password = None
        self._psk_identity = None
        self._psk_key = None
        self._client = None

        self._emit_connected.connect(self.connected, Qt.QueuedConnection)
        self._emit_disconnected.connect(self.disconnected, Qt.QueuedConnection)
        self._emit_error.connect(self.error_occurred, Qt.QueuedConnection)
        self._emit_message.connect(self._dispatch_message, Qt.QueuedConnection)

    def configure(self, host="localhost", port=1883, bbs_id=None,
                  username=None, password=None,
                  psk_identity=None, psk_key=None):
        self._host = host
        self._port = port
        self._bbs_id = bbs_id
        self._username = username
        self._password = password
        self._psk_identity = psk_identity
        self._psk_key = psk_key

    def connect_to_broker(self):
        if self._client is not None:
            self.disconnect_from_broker()

        self._client = paho_mqtt.Client(
            paho_mqtt.CallbackAPIVersion.VERSION2,
            protocol=paho_mqtt.MQTTv5,
        )

        if self._psk_identity and self._psk_key:
            ctx = make_psk_context(self._psk_identity, self._psk_key)
            self._client.tls_set_context(ctx)
            self._client.tls_insecure_set(True)

        if self._username:
            self._client.username_pw_set(self._username, self._password)
        elif self._password and self._psk_identity:
            self._client.username_pw_set(self._psk_identity, self._password)

        self._client.on_connect = self._on_connect
        self._client.on_disconnect = self._on_disconnect
        self._client.on_message = self._on_message
        self._client.on_connect_fail = self._on_connect_fail
        self._client.reconnect_delay_set(min_delay=5, max_delay=60)
        self._client.enable_logger()

        try:
            self._client.connect_async(self._host, self._port)
            self._client.loop_start()
        except Exception as e:
            self._emit_error.emit(f"Connect error: {e}")

    def disconnect_from_broker(self):
        if self._client is not None:
            self._client.loop_stop()
            self._client.disconnect()
            self._client = None

    def _topic_prefix(self):
        if self._bbs_id:
            return f"sbbs/{self._bbs_id}"
        return "sbbs/+"

    def publish(self, topic_suffix, payload):
        if self._client is None or not self._bbs_id:
            return
        topic = f"sbbs/{self._bbs_id}/{topic_suffix}"
        data = payload.encode() if isinstance(payload, str) else payload
        self._client.publish(topic, data)

    def recycle_server(self, server):
        self.publish(f"host/+/server/{server}/recycle", "")

    def pause_server(self, server):
        self.publish(f"host/+/server/{server}/pause", "")

    def resume_server(self, server):
        self.publish(f"host/+/server/{server}/resume", "")

    def clear_server(self, server):
        self.publish(f"host/+/server/{server}/clear", "")

    def trigger_event(self, code):
        self.publish("exec", code)

    def trigger_callout(self, hub_id):
        self.publish("call", hub_id)

    def send_node_message(self, node_num, message):
        self.publish(f"node/{node_num}/msg", message)

    def set_node(self, node_num, prop, value):
        self.publish(f"node/{node_num}/set/{prop}", str(value))

    # -- paho callbacks (called from paho's network thread) --

    def _on_connect(self, client, userdata, flags, reason_code, properties=None):
        if reason_code == 0:
            prefix = self._topic_prefix()
            subs = [
                f"{prefix}/host/+/server/+/log/#",
                f"{prefix}/log/#",
                f"{prefix}/node/#",
                f"{prefix}/host/+/server/+/client/action/#",
                f"{prefix}/host/+/server/+/client",
                f"{prefix}/host/+/server/+",
                f"{prefix}/host/+/server/+/state/#",
                f"{prefix}/host/+/server/+/served",
                f"{prefix}/host/+/server/+/highwater",
                f"{prefix}/host/+/server/+/error_count",
                f"{prefix}/host/+/login_attempts/#",
            ]
            for topic in subs:
                client.subscribe(topic)
            self._emit_connected.emit()
        else:
            self._emit_error.emit(
                f"MQTT connect refused: {reason_code} ({reason_code.getName()})"
                if hasattr(reason_code, "getName")
                else f"MQTT connect refused: {reason_code}"
            )

    def _on_connect_fail(self, client, userdata):
        host = getattr(client, '_host', self._host)
        port = getattr(client, '_port', self._port)
        self._emit_error.emit(
            f"MQTT connection failed to {host}:{port} "
            f"(check hostname, port, and network)"
        )

    def _on_disconnect(self, client, userdata, flags, reason_code, properties=None):
        if reason_code != 0:
            self._emit_error.emit(f"MQTT disconnected unexpectedly: {reason_code}")
        self._emit_disconnected.emit()

    def _on_message(self, client, userdata, msg):
        self._emit_message.emit(msg.topic, msg.payload)

    # -- Qt-thread message dispatch --

    @Slot(str, bytes)
    def _dispatch_message(self, topic, payload):
        parts = topic.split("/")
        if len(parts) >= 2 and parts[0] == "sbbs" and not self._bbs_id:
            self._bbs_id = parts[1]

        text = payload.decode("utf-8", errors="replace")

        # sbbs/{id}/host/{h}/server/{srv}/log/{level}
        if len(parts) >= 8 and parts[4] == "server" and parts[6] == "log":
            server = parts[5]
            level = self._parse_log_level(parts[-1])
            timestamp, msg_text = self._split_tsv_payload(text)
            self.log_message.emit(server, level, timestamp, msg_text)
            return

        # sbbs/{id}/log/{level}
        if len(parts) >= 4 and parts[2] == "log":
            level = self._parse_log_level(parts[-1])
            timestamp, msg_text = self._split_tsv_payload(text)
            self.log_message.emit("bbs", level, timestamp, msg_text)
            return

        # sbbs/{id}/node/{n}/status (structured)
        if len(parts) >= 5 and parts[2] == "node" and parts[4] == "status":
            try:
                node_num = int(parts[3])
            except ValueError:
                return
            fields = text.split("\t")
            names = ["status", "action", "user", "connection",
                     "misc", "aux", "extaux", "errors"]
            self.node_status.emit(node_num, dict(zip(names, fields)))
            return

        # sbbs/{id}/node/{n} (verbose, human-readable)
        if len(parts) == 4 and parts[2] == "node":
            try:
                node_num = int(parts[3])
            except ValueError:
                return
            self.node_verbose.emit(node_num, text)
            return

        # sbbs/{id}/host/{h}/server/{srv}/client/action/{act}
        if (len(parts) >= 9 and parts[4] == "server"
                and parts[6] == "client" and parts[7] == "action"):
            server = parts[5]
            action = parts[8]
            fields = text.split("\t")
            names = ["timestamp", "protocol", "usernum", "username",
                     "ip", "hostname", "port", "socket"]
            self.client_update.emit(server, action, dict(zip(names, fields)))
            return

        # sbbs/{id}/host/{h}/server/{srv}/state/{state} (ignored — use server-level topic)
        if len(parts) >= 8 and parts[4] == "server" and parts[6] == "state":
            return

        # sbbs/{id}/host/{h}/login_attempts/{ip}
        if len(parts) >= 6 and parts[4] == "login_attempts":
            ip = "/".join(parts[5:])
            if not text:
                self.login_attempt.emit(ip, "clear", {})
            else:
                fields = text.split("\t")
                names = ["first", "last", "count", "dupes", "protocol", "username"]
                self.login_attempt.emit(ip, "update", dict(zip(names, fields)))
            return

        # sbbs/{id}/host/{h}/server/{srv} (server-level status, retained)
        if len(parts) == 6 and parts[4] == "server":
            state = text.split("\t", 1)[0]
            self.server_state.emit(parts[5], state)
            return

        # sbbs/{id}/host/{h}/server/{srv}/client (count)
        if (len(parts) == 7 and parts[4] == "server" and parts[6] == "client"):
            self.server_stat.emit(parts[5], "clients", text)
            return

        # sbbs/{id}/host/{h}/server/{srv}/{stat} (served/highwater/error_count)
        if (len(parts) == 7 and parts[4] == "server"
                and parts[6] in ("served", "highwater", "error_count")):
            self.server_stat.emit(parts[5], parts[6], text)
            return

    def _parse_log_level(self, level_str):
        try:
            return int(level_str)
        except ValueError:
            pass
        level_map = {
            "emergency": 0, "alert": 1, "critical": 2, "error": 3,
            "warn": 4, "warning": 4, "notice": 5, "info": 6, "debug": 7,
        }
        return level_map.get(level_str, 6)

    def _split_tsv_payload(self, text):
        fields = text.split("\t", 1)
        if len(fields) > 1:
            return fields[0], fields[1]
        return "", text

File: src/sbbs3/qtmonitor/mqttclient.cpp
#include "mqttclient.h"
#include <QDateTime>
#include <QMqttSubscription>
#include <QTimer>
#include <QMqttTopicFilter>
#include <QSslConfiguration>
#include <QSslCipher>
#include <QSslCertificate>
#include <QSslKey>
#include <QFile>

MqttClient::MqttClient(QObject *parent)
	: QObject(parent)
	, m_client(nullptr)
	, m_socket(nullptr)
{
}

void MqttClient::initClient()
{
	m_client = new QMqttClient(this);
	m_client->setProtocolVersion(QMqttClient::MQTT_5_0);
	connect(m_client, &QMqttClient::connected, this, &MqttClient::onConnected);
	connect(m_client, &QMqttClient::disconnected, this, &MqttClient::onDisconnected);
	connect(m_client, &QMqttClient::errorChanged, this, &MqttClient::onErrorChanged);
}

void MqttClient::configure(const QString &host, quint16 port,
                           const QString &bbsId,
                           const QString &username, const QString &password,
                           const QByteArray &pskIdentity, const QByteArray &pskKey,
                           const QString &caFile, const QString &certFile,
                           const QString &keyFile)
{
	m_host = host;
	m_port = port;
	m_bbsId = bbsId.isEmpty() ? QString() : bbsId;
	m_username = username;
	m_password = password;
	m_pskIdentity = pskIdentity;
	m_pskKey = pskKey;
	m_caFile = caFile;
	m_certFile = certFile;
	m_keyFile = keyFile;
}

void MqttClient::connectToBroker()
{
	disconnectFromBroker();
	initClient();

	if (!m_username.isEmpty()) {
		m_client->setUsername(m_username);
		m_client->setPassword(m_password);
	} else if (!m_password.isEmpty() && !m_pskIdentity.isEmpty()) {
		m_client->setUsername(QString::fromUtf8(m_pskIdentity));
		m_client->setPassword(m_password);
	}

	bool usePsk = !m_pskIdentity.isEmpty() && !m_pskKey.isEmpty();
	bool useTls = usePsk || !m_caFile.isEmpty();

	if (useTls) {
		m_socket = new QSslSocket(this);
		if (usePsk)
			connect(m_socket, &QSslSocket::preSharedKeyAuthenticationRequired,
			        this, &MqttClient::onPskRequired);
		if (!m_ignoredSslErrors.isEmpty())
			m_socket->ignoreSslErrors(m_ignoredSslErrors);
		connect(m_socket, &QSslSocket::sslErrors, this, [this](const QList<QSslError> &errors) {
			m_socket->ignoreSslErrors(errors);
			QList<QSslError> newErrors;
			for (const auto &e : errors)
				if (!m_ignoredSslErrors.contains(e))
					newErrors << e;
			if (newErrors.isEmpty())
				return;
			QStringList msgs;
			for (const auto &e : newErrors)
				msgs << e.errorString();
			QMetaObject::invokeMethod(this, [this, newErrors, msgs] {
				emit sslErrorsOccurred(newErrors, msgs);
			}, Qt::QueuedConnection);
		});
		connect(m_socket, &QAbstractSocket::errorOccurred, this, [this] {
			emit errorOccurred("Socket error: " + m_socket->errorString());
		});
		connect(m_socket, &QSslSocket::encrypted, this, [this] {
			m_client->setTransport(m_socket, QMqttClient::IODevice);
			m_client->setHostname(m_host);
			m_client->setPort(m_port);
			m_client->connectToHost();
		});

		QSslConfiguration config = m_socket->sslConfiguration();

		if (!m_caFile.isEmpty()) {
			QList<QSslCertificate> caCerts = QSslCertificate::fromPath(m_caFile);
			if (!caCerts.isEmpty()) {
				config.setCaCertificates(caCerts);
				config.setPeerVerifyMode(QSslSocket::VerifyPeer);
			} else {
				emit errorOccurred("Failed to load CA certificate: " + m_caFile);
				config.setPeerVerifyMode(QSslSocket::VerifyNone);
			}
		} else {
			config.setPeerVerifyMode(QSslSocket::VerifyNone);
		}

		if (!m_certFile.isEmpty() && !m_keyFile.isEmpty()) {
			QList<QSslCertificate> certs = QSslCertificate::fromPath(m_certFile);
			if (!certs.isEmpty())
				config.setLocalCertificate(certs.first());
			else
				emit errorOccurred("Failed to load client certificate: " + m_certFile);

			QFile keyFile(m_keyFile);
			if (keyFile.open(QIODevice::ReadOnly)) {
				QSslKey key(&keyFile, QSsl::Rsa, QSsl::Pem);
				if (key.isNull()) {
					keyFile.seek(0);
					key = QSslKey(&keyFile, QSsl::Ec, QSsl::Pem);
				}
				if (!key.isNull())
					config.setPrivateKey(key);
				else
					emit errorOccurred("Failed to load private key: " + m_keyFile);
			} else {
				emit errorOccurred("Cannot open key file: " + m_keyFile);
			}
		}

		if (usePsk) {
			config.setProtocol(QSsl::TlsV1_2);
			QList<QSslCipher> pskCiphers;
			for (const auto &c : QSslConfiguration::supportedCiphers())
				if (c.name().contains("PSK"))
					pskCiphers << c;
			if (!pskCiphers.isEmpty())
				config.setCiphers(pskCiphers);
		}

		m_socket->setSslConfiguration(config);
		m_socket->connectToHostEncrypted(m_host, m_port);
	} else {
		m_client->setHostname(m_host);
		m_client->setPort(m_port);
		m_client->connectToHost();
	}
}

void MqttClient::disconnectFromBroker()
{
	if (m_client) {
		if (m_client->state() != QMqttClient::Disconnected)
			m_client->disconnectFromHost();
		delete m_client;
		m_client = nullptr;
	}
	if (m_socket) {
		m_socket->abort();
		delete m_socket;
		m_socket = nullptr;
	}
}

void MqttClient::ignoreSslErrors(const QList<QSslError> &errors)
{
	m_ignoredSslErrors += errors;
}

void MqttClient::onPskRequired(QSslPreSharedKeyAuthenticator *auth)
{
	auth->setIdentity(m_pskIdentity);
	auth->setPreSharedKey(m_pskKey);
}

void MqttClient::onConnected()
{
	QString prefix = topicPrefix();
	const QStringList topics = {
		prefix + "/host/+/server/+/log/#",
		prefix + "/log/#",
		prefix + "/node/#",
		prefix + "/host/+/server/+/client/action/#",
		prefix + "/host/+/server/+/client",
		prefix + "/host/+/server/+",
		prefix + "/host/+/server/+/version",
		prefix + "/host/+/server/+/served",
		prefix + "/host/+/server/+/highwater",
		prefix + "/host/+/server/+/error_count",
		prefix + "/action/#",
		prefix + "/host/+/event/log/#",
		prefix + "/host/+/login_attempts/#",
		prefix + "/host/+/server/+/max_concurrent/#",
		prefix,
	};
	for (const auto &t : topics) {
		auto *sub = m_client->subscribe(QMqttTopicFilter(t));
		if (sub)
			connect(sub, &QMqttSubscription::messageReceived,
			        this, &MqttClient::onSubscriptionMessage);
	}
	auto *sysSub = m_client->subscribe(QMqttTopicFilter("$SYS/#"));
	if (sysSub)
		connect(sysSub, &QMqttSubscription::messageReceived,
		        this, &MqttClient::onSubscriptionMessage);

	QString clientListFilter = prefix + "/host/+/server/+/client/list";
	auto *clientListSub = m_client->subscribe(QMqttTopicFilter(clientListFilter));
	if (clientListSub) {
		connect(clientListSub, &QMqttSubscription::messageReceived,
		        this, &MqttClient::onSubscriptionMessage);
		QTimer::singleShot(2000, this, [this, clientListFilter] {
			if (m_client)
				m_client->unsubscribe(QMqttTopicFilter(clientListFilter));
		});
	}

	emit connected();
}

void MqttClient::onDisconnected()
{
	emit disconnected();
}

void MqttClient::onErrorChanged(QMqttClient::ClientError error)
{
	if (error != QMqttClient::NoError)
		emit errorOccurred(QStringLiteral("MQTT error: %1").arg(static_cast<int>(error)));
}

void MqttClient::onSubscriptionMessage(const QMqttMessage &msg)
{
	QHash<QString, QString> userProps;
	const auto props = msg.publishProperties().userProperties();
	for (const auto &p : props)
		userProps.insert(p.name(), p.value());
	dispatchMessage(msg.topic().name(), QString::fromUtf8(msg.payload()), userProps);
}

bool MqttClient::isConnected() const
{
	return m_client && m_client->state() == QMqttClient::Connected;
}

void MqttClient::publish(const QString &topicSuffix, const QByteArray &payload)
{
	if (!m_bbsId.isEmpty() && isConnected())
		m_client->publish(QMqttTopicName("sbbs/" + m_bbsId + "/" + topicSuffix), payload);
}

QString MqttClient::topicPrefix() const
{
	return m_bbsId.isEmpty() ? QStringLiteral("sbbs/+") : QStringLiteral("sbbs/") + m_bbsId;
}

void MqttClient::recycleHost(const QString &host)                          { publish("host/" + host + "/recycle", {}); }
void MqttClient::pauseHost(const QString &host)                            { publish("host/" + host + "/pause", {}); }
void MqttClient::resumeHost(const QString &host)                           { publish("host/" + host + "/resume", {}); }
void MqttClient::clearHost(const QString &host)                            { publish("host/" + host + "/clear", {}); }
void MqttClient::recycleServer(const QString &host, const QString &server) { publish("host/" + host + "/server/" + server + "/recycle", {}); }
void MqttClient::pauseServer(const QString &host, const QString &server)   { publish("host/" + host + "/server/" + server + "/pause", {}); }
void MqttClient::resumeServer(const QString &host, const QString &server)  { publish("host/" + host + "/server/" + server + "/resume", {}); }
void MqttClient::clearServer(const QString &host, const QString &server)    { publish("host/" + host + "/server/" + server + "/clear", {}); }
void MqttClient::clearLoginAttempt(const QString &host, const QString &ip)  { publish("host/" + host + "/clear", ip.toUtf8()); }
void MqttClient::triggerEvent(const QString &code)     { publish("exec", code.toUtf8()); }
void MqttClient::triggerCallout(const QString &hubId)  { publish("call", hubId.toUtf8()); }
void MqttClient::setNode(int n, const QString &prop, const QString &val) { publish(QStringLiteral("node/%1/set/%2").arg(n).arg(prop), val.toUtf8()); }
void MqttClient::sendNodeMessage(int n, const QString &msg) { publish(QStringLiteral("node/%1/msg").arg(n), msg.toUtf8()); }

static QString formatTimestamp(const QString &raw)
{
	if (raw.isEmpty())
		return raw;
	QDateTime dt = QDateTime::fromString(raw, Qt::ISODate);
	if (!dt.isValid())
		dt = QDateTime::fromString(raw.left(15), "yyyyMMdd'T'HHmmss");
	return dt.isValid() ? dt.toString("MMM dd hh:mm:ss") : raw;
}

void MqttClient::dispatchMessage(const QString &topic, const QString &text,
                                  const QHash<QString, QString> &userProps)
{
	QStringList parts = topic.split('/');

	// $SYS/broker/version
	if (topic == "$SYS/broker/version") {
		emit brokerVersion(text);
		return;
	}
	// $SYS/broker/log/{level}
	if (parts.size() == 4 && parts[0] == "$SYS" && parts[1] == "broker" && parts[2] == "log") {
		auto [timestamp, msg] = splitTsvPayload(text, userProps);
		emit brokerLog(parseLogLevel(parts[3]), timestamp, msg);
		return;
	}

	if (parts.size() >= 2 && parts[0] == "sbbs" && m_bbsId.isEmpty())
		m_bbsId = parts[1];

	// sbbs/{id} — BBS system name
	if (parts.size() == 2 && parts[0] == "sbbs" && !text.isEmpty()) {
		emit bbsName(text);
		return;
	}

	if (parts.size() >= 4 && parts[2] == "host" && !m_hosts.contains(parts[3])) {
		m_hosts.append(parts[3]);
		emit hostDiscovered(parts[3]);
	}

	QString host = (parts.size() >= 4 && parts[2] == "host") ? parts[3] : QString();

	// sbbs/{id}/host/{h}/server/{srv}/log/{level}
	if (parts.size() >= 8 && parts[4] == "server" && parts[6] == "log") {
		auto [timestamp, msg] = splitTsvPayload(text, userProps);
		emit logMessage(host, parts[5], parseLogLevel(parts.last()), timestamp, msg);
		return;
	}

	// sbbs/{id}/action/{type}/{detail...} — payload is timestamp\tdata
	if (parts.size() >= 4 && parts[2] == "action") {
		QString timestamp, payload;
		auto it = userProps.constFind(QStringLiteral("time"));
		if (it != userProps.constEnd()) {
			timestamp = formatTimestamp(it.value());
			payload = text;
		} else {
			int tab = text.indexOf('\t');
			if (tab >= 0) {
				timestamp = formatTimestamp(text.left(tab));
				payload = text.mid(tab + 1);
			} else {
				payload = text;
			}
		}
		QString action = parts[3];
		QString detail = QStringList(parts.mid(4)).join('/');
		emit bbsAction(action, detail, timestamp, payload);
		return;
	}

	// sbbs/{id}/host/{h}/event/log/{level}
	if (parts.size() >= 7 && parts[4] == "event" && parts[5] == "log") {
		auto [timestamp, msg] = splitTsvPayload(text, userProps);
		emit eventLogMessage(host, parseLogLevel(parts.last()), timestamp, msg);
		return;
	}

	// sbbs/{id}/log/{level}
	if (parts.size() >= 4 && parts[2] == "log") {
		auto [timestamp, msg] = splitTsvPayload(text, userProps);
		emit logMessage(host, "bbs", parseLogLevel(parts.last()), timestamp, msg);
		return;
	}

	// sbbs/{id}/node/{n}/status (structured)
	if (parts.size() >= 5 && parts[2] == "node" && parts[4] == "status") {
		bool ok;
		int nodeNum = parts[3].toInt(&ok);
		if (!ok) return;
		QStringList fields = text.split('\t');
		QStringList names = {"status", "action", "user", "connection", "misc", "aux", "extaux", "errors"};
		QVariantMap data;
		for (int i = 0; i < qMin(fields.size(), names.size()); ++i)
			data[names[i]] = fields[i];
		emit nodeStatus(host, nodeNum, data);
		return;
	}

	// sbbs/{id}/node/{n} (verbose)
	if (parts.size() == 4 && parts[2] == "node") {
		bool ok;
		int nodeNum = parts[3].toInt(&ok);
		if (!ok) return;
		emit nodeVerbose(host, nodeNum, text);
		return;
	}

	// sbbs/{id}/host/{h}/server/{srv}/client/list
	if (parts.size() == 8 && parts[4] == "server" && parts[6] == "client" && parts[7] == "list") {
		QStringList names = {"timestamp", "protocol", "usernum", "username", "ip", "hostname", "port", "socket"};
		for (const auto &line : text.split('\n', Qt::SkipEmptyParts)) {
			QStringList fields = line.split('\t');
			QVariantMap data;
			for (int i = 0; i < qMin(fields.size(), names.size()); ++i)
				data[names[i]] = fields[i];
			emit clientUpdate(host, parts[5], "connect", data);
		}
		return;
	}

	// sbbs/{id}/host/{h}/server/{srv}/client/action/{act}
	if (parts.size() >= 9 && parts[4] == "server" && parts[6] == "client" && parts[7] == "action") {
		QStringList fields = text.split('\t');
		QStringList names = {"timestamp", "protocol", "usernum", "username", "ip", "hostname", "port", "socket"};
		QVariantMap data;
		for (int i = 0; i < qMin(fields.size(), names.size()); ++i)
			data[names[i]] = fields[i];
		emit clientUpdate(host, parts[5], parts[8], data);
		return;
	}

	// sbbs/{id}/host/{h}/server/{srv}/state/{state} — ignored, use server-level topic
	if (parts.size() >= 8 && parts[4] == "server" && parts[6] == "state")
		return;

	// sbbs/{id}/host/{h}/login_attempts/{ip}
	if (parts.size() >= 6 && parts[4] == "login_attempts") {
		QString ip = QStringList(parts.mid(5)).join('/');
		if (text.isEmpty()) {
			emit loginAttempt(host, ip, "clear", {});
		} else {
			QStringList fields = text.split('\t');
			QStringList names = {"first", "last", "count", "dupes", "protocol", "username"};
			QVariantMap data;
			for (int i = 0; i < qMin(fields.size(), names.size()); ++i)
				data[names[i]] = fields[i];
			emit loginAttempt(host, ip, "update", data);
		}
		return;
	}

	// sbbs/{id}/host/{h}/server/{srv}/max_concurrent/{ip}
	if (parts.size() >= 8 && parts[4] == "server" && parts[6] == "max_concurrent") {
		QString ip = QStringList(parts.mid(7)).join('/');
		if (text.isEmpty())
			emit maxConcurrent(host, parts[5], ip, "clear", 0);
		else
			emit maxConcurrent(host, parts[5], ip, "update", text.toInt());
		return;
	}

	// sbbs/{id}/host/{h}/server/{srv}/client (count)
	if (parts.size() == 7 && parts[4] == "server" && parts[6] == "client") {
		emit serverStat(host, parts[5], "clients", text);
		return;
	}

	// sbbs/{id}/host/{h}/server/{srv} (server-level status)
	if (parts.size() == 6 && parts[4] == "server") {
		emit serverState(host, parts[5], text.split('\t').first());
		return;
	}

	// sbbs/{id}/host/{h}/server/{srv}/version
	if (parts.size() == 7 && parts[4] == "server" && parts[6] == "version") {
		emit serverVersion(host, parts[5], text);
		return;
	}

	// sbbs/{id}/host/{h}/server/{srv}/{stat}
	if (parts.size() == 7 && parts[4] == "server"
	    && (parts[6] == "served" || parts[6] == "highwater" || parts[6] == "error_count")) {
		emit serverStat(host, parts[5], parts[6], text);
		return;
	}
}

int MqttClient::parseLogLevel(const QString &level) const
{
	bool ok;
	int n = level.toInt(&ok);
	if (ok) return n;
	static const QHash<QString, int> map = {
		{"emergency", 0}, {"alert", 1}, {"critical", 2}, {"error", 3},
		{"warn", 4}, {"warning", 4}, {"notice", 5}, {"info", 6}, {"debug", 7},
	};
	return map.value(level, 6);
}

QPair<QString, QString> MqttClient::splitTsvPayload(const QString &text,
                                                     const QHash<QString, QString> &userProps) const
{
	auto it = userProps.constFind(QStringLiteral("time"));
	if (it != userProps.constEnd())
		return {formatTimestamp(it.value()), text};
	return {{}, text};
}
